### ISSUE TITLE
Neutron L3 Router Monitoring

### ISSUE THREAD

**Issue by mrhillsman** [rcbops/rpc-openstack#460]
Related to customer request for monitoring of Neutron L3 router in order to assist with resolving network outage issue that could have been resolved faster with an alert.

---
###### For each pull request, update the branch line to reference it.
- [ ] master
- [ ] mitaka-13.0
- [ ] liberty-12.2
- [ ] kilo
##### Upgrade impact. Consider the impact on moving between major, minor and patch releases.
- [ ] This change does not impact upgrades
- [ ] This change addresses a potential upgrade impact
- [ ] The upgrade impact is being tracked by another issue.

The issue tracking the upgrade impact is:
##### Testing impact
- [ ] An issue/pull-request has been create or exists to track the testing gap, and is referenced below
- [ ] This change includes additional testing
- [ ] This change is tested by the existing setup

Additional testing links:
##### Release note
- [ ] This change includes a release note.
- [ ] This change does not require a release note.
##### Does your change require documentation?
- [ ] Yes, and I submitted an issue to [Docs issues](https://github.com/rackerlabs/docs-rpc/issues)
- [ ] Yes, and I contributed a documentation patch to [docs-rpc](https://github.com/rackerlabs/docs-rpc)
- [ ] No documentation required for this change

Additional documentation links (please add links if you have ticked yes):

**Comment by claco** [comment]
It seems like we're adding a new plugin, but we're not also adding checks/alarms?

**Comment by BjoernT** [comment]
We do not need in particular checks. Whenever we exit this script with a >0 return code, it'll generate a alarm anyway

**Comment by prometheanfire** [comment]
@stevelle you said you'd do the check, just making a note here

**Comment by stevelle** [comment]
Without a check, this script wont be run, so it needs something more. Also a squash.
With the pending work to refactoring for maas (https://github.com/rcbops/rpc-openstack/tree/refactor/maas-templates) the method of scheduling a check will be much simpler. 

I will put together a template to be added to this issue that will work with the refactor to schedule a check for this.

**Comment by mancdaz** [comment]
@hughsaunders looks like he is going to pick up the PR

**Comment by JCallicoat** [comment]
The PR was closed. We still need a solution here.

**Comment by claco** [comment]
@JCallicoat @BjoernT Refresh my memory. Was this the one discussed previously that needed some magic determining how to monitor given if things are in, or out of a Router in Neutron?

**Comment by BjoernT** [comment]
correct since you can't always ping the gateways from the br-mgmt/bond0 network

**Comment by BjoernT** [comment]
Yeah and then there was plumgrid... 
It's a tough question, the plugin needs to be smart enough to deal with linuxridge agents, plumgrid and eventually OVS.
For linuxbridge agent I would see the following pseudo code as sufficient:

- Get router id : neutron router-list
- Alert if admin_state_up but not alive
- Determine active agent for a network : neutron l3-agent-list-hosting-router <id>
- Alert if fails: lxc-attach -n $(lxc-ls |grep neutron_agents_container |tail -1) ip netns exec qrouter-<router-id> ping -c2 8.8.8.8
- Get a list of active floating ips, fixed_ips and floating_ip_address need to be set: neutron floatingip-list  | awk -F \|  '/([0-9a-z]+-)+/ { print $3 ":" $4 }'
- lxc-attach -n $(lxc-ls |grep neutron_agents_container |tail -1) ip netns exec qrouter-<router-id> iptables -t nat -L neutron-l3-agent-PREROUTING -n |grep DNAT > floats
- Search for the active floating ips inside the floats file

**Comment by claco** [comment]
Cool! Thanks @BjoernT

**Comment by alextricity25** [comment]
I would like to take a stab at this using https://github.com/hughsaunders/rpc-openstack/commit/47b2226ee2ee33a8ccf27ff0a0e171eaec63d5f3 as a starting point.
What do you think @claco?

**Comment by b3rnard0** [comment]
@mancdaz @claco Is this a but or an enhancement? If the latter, what's the size to determine when we should put this into our approved list.

**Comment by mancdaz** [comment]
This is an enhancement. Additional bits of monitoring I would suggest are type-enhancement-small

**Comment by Apsu** [comment]
@BjoernT @JCallicoat I'd like to try to move this forward, but I want to nail down criteria and methods in hopes of covering linux-bridge, OVS, Plumgrid, and whatever else.

First of all, I see two separate considerations here:

* Verify that routers are scheduled to (at least one) agent, admin_state is up, and are actually "there"
* Test that routers are actually routing/floats are actually on interfaces

The first is very in-line with the existing types of checks as I understand it, while the second is more akin to functional testing or QC than basic monitoring, even if we automate it. @hughsaunders took a good stab at covering all of this with hughsaunders@47b2226 as @alextricity25 mentioned, but I'd like to get some more explicit requirements for what exactly you want monitored, and how critical each aspect is to resolving the issue.

Hopefully we can figure out an approach for each requirement that might be adaptable to other L2 plugin systems. We also need to know if/how this interacts with L3HA/DVR considerations, as those systems affect how/where you look and check.

**Comment by Apsu** [comment]
Waiting for feedback

**Comment by BjoernT** [comment]
@Apsu see https://github.com/rcbops/rpc-openstack/issues/460#issuecomment-195406808 in the end we want to check few things:
- Router is up
- Router if function for both ingress and egress traffic
- Floating IPs are working

**Comment by Apsu** [comment]
(Pasting from email thread)

We just had a meeting on this, here was the agenda just so everyone has it in the thread:

—

Just wanted to get together to discuss these requirements in a bit more depth.

Things I’d like to cover:

* What exact issues are we trying to solve/prevent from reoccuring?
* Are there different levels of monitoring with different priorities so we can split the work appropriately?
* Details on exactly what we intend to monitor and if/how it relates to SLAs so we can evaluate methods
* Timeframes for completion for prioritization

—

And the meeting notes I took:

—

- Linux Bridge only for now
- Long-term extended to other L2 plugins
- LB
 - Ingress
   - ARPing from router netns to one host per inside network
     - Minimize traffic saturation and ARP cache impact
   - Floats are on external interface (priority)
   - DNAT/SNAT rules are in iptables (priority)
 - Egress
   - Default route is in routing table (priority)
   - Gateway of default route is reachable (priority)
   - Inside interfaces are up and have appropriate IPs
     - Generally gateway IPs of internal networks
   - Optional external IP for reachability
     - Only if appropriate, such as 8.8.8.8 for internet transit tests
- SLA
 - 15 minute end to end, including RBA/CORE
 - 5 minutes for alert to be in-line with hatool failover delay
   - Different thresholds and samples for different criteria
     - Unscheduled router (priority)
     - Gateway connectivity (priority)
     - Inside network connectivity
     - TBD

—

I think this gives us a place to start and I will work on getting these requirements and acceptance criteria into card form.

**Comment by b3rnard0** [comment]
Is this still block based on your conversation from yesterday?

**Comment by Apsu** [comment]
@b3rnard0 No, this is now unblocked I believe, and ready to be checklisted and potentially split into other cards for prioritization.

**Comment by Apsu** [comment]
Got stuck in MaaS and code review/PR/design mines, will split this out tomorrow.

**Comment by Apsu** [comment]
Monitoring requirements and acceptance criteria
---

---

These tests are to be performed outside the router namespaces:

### Scheduling

- [ ] Verify routers are scheduled to an L3 agent and are active/admin_state up

---

These tests are to be performed inside each router namespace:

### Ingress

- [ ] Verify assigned floating IPs are actually on external interface
- [ ] Verify DNAT/SNAT rules for floating IPs are present in iptables
- [ ] ARPing to one instance per inside network (low-priority)

### Egress

- [ ] Verify default route is in routing table
- [ ] Verify nexthop of default route is reachable via ICMP
- [ ] Verify inside interfaces are UP and correctly configured with inside network gateway IPs
- [ ] Verify external IP is reachable via ICMP (optional, low-priority)

---

These considerations concern the MaaS plugin behavior:

### SLA

- [ ] Consider tests failed no more often than every 5 minutes to not 'beat' the hatool failover max delay
- [ ] Figure out thresholds and sample criteria based on specific things being tested
    - Unscheduled router
    - Gateway connectivity
    - Inside network connectivity (low-priority/larger sample count?)

**Comment by git-harry** [comment]
Requirements gathered, pending available person to complete code.

**Comment by Apsu** [comment]
Might be available to make some progress on this tomorrow.

**Comment by tomc603** [comment]
@git-harry @Apsu Would it make sense to close this particular card now that requirements are gathered, and create one or more cards to implement the work necessary for monitoring Neutron properly?

**Comment by Apsu** [comment]
@tomc603 Yep, that's basically what I had in mind to begin moving forward on this, especially so that there's not this one card sitting in Doing being blocked forever, and we can have backlogged cards to get prioritized and doled out. If you're volunteering to help, I wouldn't turn you down :D

**Comment by Apsu** [comment]
Plugin development is in progress.

**Comment by Apsu** [comment]
Still in progress, was diverted to 1440 pretty much all of today.

**Comment by Apsu** [comment]
Update time!

Still doing some research and a bit of hacking but because MaaS plugins don't run inside containers, my current approach is to reach into the neutron-agents containers by accessing their LXC namespaces, in order to get to their network namespaces. I'm currently making use of the pyroute2 pip package in my test code, which gives me a Netlink interface to easily probe and manipulate interfaces and network namespaces.

Once I finish testing the strategy, I'll integrate basic code into the MaaS plugin and update/link to my branch.

**Comment by Apsu** [comment]
May not get the code integrated into a MaaS plugin by EOD, so here's a gist of my progress so far: https://gist.github.com/Apsu/6c4e21f1d17c7b4e096b039481d6ccd9

Essentially using the lxc and pyroute2 modules to iterate over neutron_agent containers, enter their namespaces, get the list of network namespaces inside, organize into DHCP/Router, iterate over them, get their interface names/states/IPs, with some filtering and access helpers.

Next step is to pull this into a plugin skeleton with neutronclient to get agent/router/network/subnet/port info for the static checks the requirements call for, and use pyroute2's NSPopen wrapper to run dynamic checks in the appropriate namespaces. Then we can define metrics/alarms and tie them together.

**Comment by tomc603** [comment]
@Apsu Has basic working code. He is spinning up a new AIO, due to his previous environment returning 5xx.

**Comment by Apsu** [comment]
Had some more AIO issues, will have basic integration in a branch in the morning.

**Comment by Apsu** [comment]
https://github.com/Apsu/rpc-openstack/commit/7879914d4c19e9c19ac65c96be3027cedac93226 for https://github.com/Apsu/rpc-openstack/tree/neutron-l3-monitor in place to work from. Just a few more bits to wire in then present to MaaS.

**Comment by Apsu** [comment]
Keep running into weird galera package signing issues on AIO runs, threw me off course for quite a while. But I'm in a working-cluster state, with Neutron behaving correctly, so I am moving forward with the neutronclient integration, which is essentially the last step. I'll have an EOD update and code ready to go.

**Comment by claco** [comment]
@Apsu Related maybe? https://github.com/rcbops/rpc-openstack/pull/1529

**Comment by Apsu** [comment]
@claco Definitely related. Good catch.

**Comment by Apsu** [comment]
Working through requirements, updating branch in the morning, will check the boxes

**Comment by Apsu** [comment]
Wired in router state checks

**Comment by Apsu** [comment]
Left off with a lot of time spent trying to predict the pattern of Neutron's iptables rules for floats and routers and whether or not I could identify if they were present/missing well enough programmatically. I'm using an iptables/netfilter library to parse the rules, so that helps substantially, but I think I'm going to move those requirements to low priority for now and revisit once the rest are in place.

**Comment by Apsu** [comment]
Didn't make a ton of progress today due to illness and escalation.  Will be focusing on the easy requirements tomorrow to bang out as much as possible.

**Comment by Apsu** [comment]
* Updated requirements list to deprioritize iptables rules verification
* Rebuilt AIO with newer SHA checkout so deployment doesn't fail with weird package issues
* Updated branch with rebase of upstream master to incorporate changes

**Comment by claco** [comment]
Marking as blocked for assignee due to prio-expedite: https://waffle.io/rcbops/u-suk-dev/cards/581cf18c44a2c4950003db54

Can be taken over by someone else, or will be finished post expedite.

**Comment by Apsu** [comment]
This has been unblocked, so I'm back into it now. Got reoriented, spun up a new AIO and got started on the rest of the work to be done. Will link to a branch update once I finish up a few things and get them pushed too.

**Comment by Apsu** [comment]
Holidays got me all screwed up, but progress yesterday was updating the AIO to current code and fixing a few issues I ran into with that process. Spent some time seeing if they were bugs or flukes. Not sure if I'll be getting assigned new work today or not, so I'm holding off for the moment until I see what's going on there.

**Comment by claco** [comment]
Please finish cards in doing, Any new cards/work from meetings this week will start after Doings are Done. :-)

**Comment by Apsu** [comment]
@claco Roger. On it, boss.

**Comment by Apsu** [comment]
Pushed update to my fork, started to organize the code and data so I can close the loop.

* Breakout more Neutron dicts
* Pass into container and netns
* Correlate IPs with router ports
* Prepare to check and return status on floats/IPs

And some other miscellaneous work to get what I need into the container and then network namespaces.

Hope to be at the point of writing alarm language tomorrow.

**Comment by Apsu** [comment]
Had some issues to deal with due to the rain, worked on this a bit over the weekend, pushing in the morning.

**Comment by Apsu** [comment]
Goal is to be done by end of week. Remaining work:

* Fixup data structures and helper functions to get floats and gateway IPs from a network bound to a router to get ports from the router bound to a device to get netns for the router that has the binding to figure out interface the IPs should be on to check off most of the remaining checkboxes.
* Learn to enjoy life again
* Tell Neutron devs about their terrible life choices and database design

**Comment by Apsu** [comment]
Ok, I've got everything wired in that I need but unfortunately I've been fighting a snag that is preventing me from getting data back out of the container. The LXC library's attach_* functions do not return anything but the status of waitpid(), which is success/error. The payload passed in is to an exec() function and is void*, which means the Python wrapper won't let me carry global variable scope or transfer data back and forth through the usual means.

I might be able to retain the stdout FD, based on looking at the C API arguments, but worst case I suppose I could write to a tempfile that the code outside the container namespace could try to read from /var/lib/lxc/$container/rootfs/$blah.

All of this is terrible. But the remaining bit of the monitoring code is easy, just been trying to figure out how to cross two namespace barriers and carry information back. I'll update tomorrow after I try some... clever approaches and see what I can make work.

**Comment by claco** [comment]
Because of the amount of work arounds, this is turning into a not so simple solution that likely won't be predictable or stable. We're going to abandon this request at this time. We have saved the current code as it is, and may revisit this in the future as part of monitoring/scaling discussions.

**Comment by BjoernT** [comment]
Reopening since we need monitoring like this, due to repeating issues with RabbitMQ and neutron router availability. We also should consider to have a l3ha compatible solution

**Comment by claco** [comment]
Re-Closing. The Unified Gating group has an entire milestone of work progressing around the support/qe identified testing gaps located https://rpc-openstack.atlassian.net/secure/RapidBoard.jspa?rapidView=5 and https://one.rackspace.com/display/PrivateCloud/Unified+Gating

Formerly https://waffle.io/rcbops/u-suk-dev?label=feature-gating,swimlane-improvements1

Still looking for that etherpad page form those meetings prior to the milestone being committed to.
